### PR TITLE
Refactoring message line

### PR DESCRIPTION
**Pull request by phong9xNcc**

### CODE REVIEW

File: libs/utils/src/lib/types/messageLine.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+export interface IMentionOnMessage {
+	user_id: string | undefined;
+	username: string | undefined;
```
Comment (anbuihoang) [inline libs/utils/src/lib/types/messageLine.ts:8]: Tạo interface dùng chung để extends, tránh trùng lặp

File: libs/components/src/lib/components/MessageWithUser/MessageLine.tsx [hunk 2]
```diff
@@ -1,54 +1,110 @@
-import { useCallback, useLayoutEffect, useState } from 'react';
-import MarkdownFormatText from '../MarkdownFormatText';
-import MessageImage from './MessageImage';
-import { useMessageLine } from './useMessageLine';
+import { selectTheme } from '@mezon/store';
```
Comment (minhlucncc) [inline libs/components/src/lib/components/MessageWithUser/MessageLine.tsx]: import { selectTheme } from '@mezon/store';
import clx from 'classnames';
import { memo, useCallback, useMemo } from 'react';
import Markdown from 'react-markdown';
import { useSelector } from 'react-redux';
import remarkGfm from 'remark-gfm';
import { ChannelHashtag, EmojiMarkdown, MentionUser } from '../../components';
import PreClass from '../MarkdownFormatText/PreClass';

interface MessageLineProps {
    line: string;
    messageId?: string;
    mode?: number;
    content: any;
}

interface RenderContentProps {
    data: any;
    mode: number;
}

const Link = memo(({ url }: { url: string }) => {
    const onClickLink = useCallback(() => {
        window.open(url, '_blank');
    }, [url]);

    return (
        <a
            style={{ color: 'rgb(59,130,246)', cursor: 'pointer' }}
            className="hover:underline"
            onClick={onClickLink}
        >
            {url}
        </a>
    );
});

const MarkdownContent = memo(({ markdown, classes }: { markdown: string, classes: string }) => (
    <article style={{ letterSpacing: '-0.01rem' }} className={classes}>
        <div className="lineText contents dark:text-white text-colorTextLightMode">
            <Markdown
                children={markdown}
                remarkPlugins={[remarkGfm]}
                components={{
                    pre: PreClass,
                    p: 'span',
                    a: (props) => (
                        <span
                            onClick={() => window.open(props.href ?? '', '_blank')}
                            rel="noopener noreferrer"
                            style={{ color: 'rgb(59,130,246)', cursor: 'pointer' }}
                            className="tagLink"
                        >
                            {props.children}
                        </span>
                    ),
                }}
            />
        </div>
    </article>
));

const RenderContent = memo(({ data, mode }: RenderContentProps) => {
    const appearanceTheme = useSelector(selectTheme);
    const { t, mentions = [], hashtags = [], emojis = [], links = [], markdowns = [] } = data;
    const elements = useMemo(
        () => [...mentions, ...hashtags, ...emojis, ...links, ...markdowns].sort((a, b) => a.start_index - b.start_index),
        [mentions, hashtags, emojis, links, markdowns]
    );

    let lastIndex = 0;
    const content = useMemo(() => {
        const tempContent: React.ReactNode[] = [];

        elements.forEach((element, index) => {
            const { start_index, end_index, channel_id, channel_lable, username, shortname, link, markdown } = element;

            if (lastIndex < start_index) {
                tempContent.push(t.slice(lastIndex, start_index));
            }

            if (channel_id && channel_lable) {
                tempContent.push(<ChannelHashtag key={`${index}${start_index}${channel_id}`} channelHastagId={`<#${channel_id}>`} />);
            }
            if (username) {
                tempContent.push(<MentionUser key={`${index}${start_index}${username}`} tagName={username} mode={mode} />);
            }
            if (shortname) {
                tempContent.push(<EmojiMarkdown key={`${index}${start_index}${shortname}`} emojiSyntax={shortname} onlyEmoji={false} />);
            }
            if (link) {
                tempContent.push(<Link key={`${index}${start_index}${link}`} url={link} />);
            }
            if (markdown) {
                const classes = clx(
                    'prose-code:text-sm prose-hr:my-0 prose-headings:my-0 prose-h1-2xl whitespace-pre-wrap prose prose-blockquote:my-0 leading-[0]',
                    { lightMode: appearanceTheme === 'light' }
                );
                tempContent.push(<MarkdownContent key={`${index}${start_index}${markdown}`} markdown={markdown} classes={classes} />);
            }
            lastIndex = end_index;
        });

        if (lastIndex < t.length) {
            tempContent.push(t.slice(lastIndex));
        }

        return tempContent;
    }, [elements, appearanceTheme, t, mode]);

    return <div>{content}</div>;
});

const MessageLine = ({ line, messageId, mode = 0, content }: MessageLineProps) => (
    <div>
        <RenderContent data={content} mode={mode} />
    </div>
);

export default memo(MessageLine);

File: libs/utils/src/lib/types/messageLine.ts (hunk 1 above)
Comment (phong9xNcc) [reply libs/utils/src/lib/types/messageLine.ts:8]: fixed

File: libs/components/src/lib/components/MessageWithUser/MessageLine.tsx (hunk 2 above)
Comment (phong9xNcc) [reply libs/components/src/lib/components/MessageWithUser/MessageLine.tsx]: updated

**Pull request merged**